docs: Add SDD for 1609 feature request (#1670)
Flank/flank: docs/feature/1609-add-apk-name-to-result-matrix.md

# Add apk name or module name to the result matrix

> Currently, the result matrix displays matrices ids only. When there are lots of additional test apks it's hard to distinguish which apk failed. Having apk file name associated with results (matrices) will give a better overview and will allow a user to find a failing apk module without the need of jumping between result URLs.

Reviewer: ```suggestion
Currently, the result matrix displays matrices ids only. When there are lots of additional test apks it's hard to distinguish which apk has failed. Having the apk file name associated with the results (matrices) will give a better overview and will allow a user to find a failing apk module without the need of jumping between result URLs.
```

│ failure │ matrix-2n50loc2ze1h4 │ NexusLowRes-29-de-portrait │ 1 test cases failed, 6 passed             │
│ success │ matrix-3eo91na6dzmd1 │ NexusLowRes-29-de-portrait │ 8 test cases passed                       │
│ failure │ matrix-1ztkep4prcq43 │ NexusLowRes-29-de-portrait │ 1 test cases failed                       │
│ failure │ matrix-w9wpp0jwt4yba │ NexusLowRes-29-de-portrait │ 23 test cases failed, 6 passed            │
│ success │ matrix-30cqwb4sc1rqh │ NexusLowRes-29-de-portrait │ 17 test cases passed                      │
│ success │ matrix-32kzu6e7a1t3h │ NexusLowRes-29-de-portrait │ 2 test cases passed                       │
│ failure │ matrix-3dp8qqblpyav7 │ NexusLowRes-29-de-portrait │ 20 test cases failed, 12 passed           │
│ failure │ matrix-zw52wsouk4zma │ NexusLowRes-29-de-portrait │ 1 test cases failed                       │
│ INVALID │ matrix-36qxkc2szybc6 │                            │ NO_TEST_RUNNER_CLASS                      │
└─────────┴──────────────────────┴────────────────────────────┴───────────────────────────────────────────┘
```

# References

- [Issue #1609](https://github.com/Flank/flank/issues/1609)

# Motivation

> Flank gives a possibility to run multiple apks in parallel within a single test run. This feature is extensively used by companies with big projects with lots of additional modules. Unfortunately, processing results in the current form is neither easy nor straightforward. The user needs to associate matrix with related apk/module which can be time-consuming and error-prone. To achieve a better result overview, flank should add apk name to the both result table and output JSON.

Reviewer: ```suggestion
Flank gives the possibility to run multiple apks in parallel within a single test run. This feature is extensively used by companies with big projects that include a multitude of additional modules. Unfortunately, processing results in the current form is neither easy nor straightforward. The user needs to associate the matrix with the related apk/modules which can be time-consuming and error-prone. To achieve a better result overview, flank should add the apk name to both the result table and the output JSON.
```

# Goals

Add an associated apk file name to the result table and output JSON.

# Design

`Flank` is not aware of any project, modules, etc so it can't find module names on its own, in a reasonable way, without breaking backward compatibility.

> On the other hand finding, the apk file name is simple and straightforward. To achieve it we need:

Reviewer: ```suggestion
On the other hand, finding the apk file name is simple and straightforward. To achieve it we need:
```

1. add new field `app` to the `SavedMatrix` data class
2. update private extension function `SavedMatrix#updateProperties` to extract apk file name from the `newMatrix`
   * find app used during tests under the `TestMatrix` -> `TestSpecification`

>    * once we have `TestSpecyfication`, iterate over all specs (android/ios/instrumentation/etc.) and find non-null

Reviewer: ```suggestion
   * once we have `TestSpecification`, iterate over all specs (android/ios/instrumentation/etc.) and find non-null
```

   * since we can run only one spec within a matrix, only one (spec) is non-null

>    * fetch `appApk` from spec (this is full `gs://` path) `TestSpecyfication#appApk`

Reviewer: ```suggestion
   * fetch `appApk` from spec (this is full `gs://` path) `TestSpecification#appApk`
```
The y in the middle here is incorrect. If the class is called this we should change it :)

Author: Nah, that my mistake, thanks!
